File: front/src/api/embedding.ts
import { withApiBase } from './base'

export interface EmbeddingSetting {
    id: number
    embedding_provider: string | null
    embedding_api_key: string | null
    embedding_base_url: string | null
    embedding_dim: number | null
    embedding_model: string | null
    embedding_max_token_size: number | null
    create_time: string
    update_time: string
    deleted: boolean
}

export interface EmbeddingSettingCreate {
    embedding_provider?: string | null
    embedding_api_key?: string | null
    embedding_base_url?: string | null
    embedding_dim?: number | null
    embedding_model?: string | null
    embedding_max_token_size?: number | null
}

export interface EmbeddingSettingUpdate {
    embedding_provider?: string | null
    embedding_api_key?: string | null
    embedding_base_url?: string | null
    embedding_dim?: number | null
    embedding_model?: string | null
    embedding_max_token_size?: number | null
}

export async function getEmbeddingSettings(skip: number = 0, limit: number = 100): Promise<EmbeddingSetting[]> {
    const response = await fetch(withApiBase(`/api/embedding-settings/?skip=${skip}&limit=${limit}`))
    if (!response.ok) {
        throw new Error('Failed to fetch embedding settings')
    }
    return response.json()
}

export async function createEmbeddingSetting(data: EmbeddingSettingCreate): Promise<EmbeddingSetting> {
    const response = await fetch(withApiBase('/api/embedding-settings/'), {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(data)
    })
    if (!response.ok) {
        throw new Error('Failed to create embedding setting')
    }
    return response.json()
}

export async function updateEmbeddingSetting(id: number, data: EmbeddingSettingUpdate): Promise<EmbeddingSetting> {
    const response = await fetch(withApiBase(`/api/embedding-settings/${id}`), {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(data)
    })
    if (!response.ok) {
        throw new Error('Failed to update embedding setting')
    }
    return response.json()
}

export interface EmbeddingTestResult {
    success: boolean
    message: string
    model?: string
    dimension?: number
}

export async function testEmbeddingConfig(config: EmbeddingSettingCreate): Promise<EmbeddingTestResult> {
    const response = await fetch(withApiBase('/api/embedding-settings/test'), {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(config),
    })
    if (!response.ok) {
        const error = await response.json().catch(() => ({ detail: 'Test failed' }))
        throw new Error(error.detail || 'Embedding test failed')
    }
    return response.json()
}

File: server/app/schemas/embedding.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime
from typing import Optional

class EmbeddingSettingBase(BaseModel):
    config_name: Optional[str] = None
    embedding_provider: Optional[str] = "openai"
    embedding_api_key: Optional[str] = None
    embedding_base_url: Optional[str] = None
    embedding_dim: Optional[int] = 1024
    embedding_model: Optional[str] = "BAAI/bge-m3"
    embedding_max_token_size: Optional[int] = 8000
    is_verified: Optional[bool] = False

class EmbeddingSettingCreate(EmbeddingSettingBase):
    pass

class EmbeddingSettingUpdate(EmbeddingSettingBase):
    embedding_provider: Optional[str] = None
    embedding_dim: Optional[int] = None
    embedding_model: Optional[str] = None
    embedding_max_token_size: Optional[int] = None

class EmbeddingSetting(EmbeddingSettingBase):
    id: int
    create_time: datetime
    update_time: datetime
    deleted: bool

    model_config = ConfigDict(from_attributes=True)
